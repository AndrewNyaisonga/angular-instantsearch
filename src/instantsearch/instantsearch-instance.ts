import { Injectable } from "@angular/core";
import instantsearch from "instantsearch.js/es";

@Injectable()
export class NgISInstance {
  public appId: string;
  public apiKey: string;
  public indexName: string;

  public numberLocale?: string;
  public searchFunction?: () => void;
  public createAlgoliaClient?: () => object;
  public searchParameters?: object | void;
  public urlSync?:
    | boolean
    | {
        mapping?: object;
        threshold?: number;
        trackedParameters?: string[];
        useHash?: boolean;
        getHistoryState?: () => object;
      };

  private instance?: {
    start: () => void;
  };

  public init(config: object) {
    this.instance = instantsearch(config);
  }

  public start() {
    this.instance.start();
  }
}
