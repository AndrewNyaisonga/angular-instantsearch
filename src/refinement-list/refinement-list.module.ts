import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";

import { NgISFooterModule } from "../footer/footer.module";
import { NgISHeaderModule } from "../header/header.module";
import { NgISRefinementList } from "./refinement-list";

@NgModule({
  declarations: [NgISRefinementList],
  entryComponents: [NgISRefinementList],
  exports: [NgISRefinementList],
  imports: [CommonModule, NgISHeaderModule, NgISFooterModule]
})
export class NgISRefinementListModule {}
