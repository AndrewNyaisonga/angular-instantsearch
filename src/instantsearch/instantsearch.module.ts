import { CommonModule } from "@angular/common";
import { ModuleWithProviders, NgModule } from "@angular/core";

import { NgISInstantSearch } from "./instantsearch";
import { NgISInstance } from "./instantsearch-instance";

@NgModule({
  declarations: [NgISInstantSearch],
  entryComponents: [NgISInstantSearch],
  exports: [NgISInstantSearch],
  imports: [CommonModule]
})
export class NgISInstantSearchModule {
  public static forRoot(): ModuleWithProviders {
    return {
      ngModule: NgISInstantSearchModule,
      providers: [NgISInstance]
    };
  }
}
