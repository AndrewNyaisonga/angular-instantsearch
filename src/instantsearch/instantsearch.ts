import { AfterViewInit, Component, Input, OnInit } from "@angular/core";

import { NgISInstance } from "./instantsearch-instance";

@Component({
  selector: "ngis-instantsearch",
  template: `<ng-content></ng-content>`
})
export class NgISInstantSearch implements AfterViewInit, OnInit {
  @Input() public config: InstantSearchConfig;

  constructor(private searchInstance: NgISInstance) {}

  public ngOnInit() {
    this.searchInstance.init(this.config);
  }

  public ngAfterViewInit() {
    this.searchInstance.start();
  }
}
