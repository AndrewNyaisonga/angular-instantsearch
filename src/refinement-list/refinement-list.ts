import { Component, Input } from "@angular/core";
import { connectRefinementList } from "instantsearch.js/es/connectors";
import { noop } from "lodash";

import { BaseWidget } from "../base-widget";
import { NgAisInstance } from "../instantsearch/instantsearch-instance";
import { bem } from "../utils";

const cx = bem("RefinementList");

@Component({
  selector: "ng-ais-refinement-list",
  template: `
    <div class="${cx()}">
      <ng-ais-header [header]="header" className="${cx(
        "header"
      )}"></ng-ais-header>

      <div class="${cx("body")}">
        <ul class="${cx("list")}">
          <li
            [ngClass]="{
              '${cx("item")}': true,
              '${cx("item", "selected")}': item.isRefined
            }"
            *ngFor="let item of state.items"
            (click)="refine($event, item)"
          >
            <label class="${cx("label")}">
              <input
                class="${cx("checkbox")}"
                type="checkbox"
                value="{{item.value}}"
                [checked]="item.isRefined"
              />
              {{item.label}}
              <span class="${cx("count")}">
                {{item.count}}
              </span>
            </label>
          </li>
        </ul>

        <button
          *ngIf="state.canToggleShowMore"
          (click)="state.toggleShowMore()"
        >
          {{state.isShowingMore ? showLessLabel : showMoreLabel}}
        </button>
      </div>

      <ng-ais-footer [footer]="footer" className="${cx(
        "footer"
      )}"></ng-ais-footer>
    </div>
  `
})
export class NgAisRefinementList extends BaseWidget {
  // render options
  @Input() public showMoreLabel: string = "Show more";
  @Input() public showLessLabel: string = "Show less";

  // connectors options
  @Input() public attributeName: string;
  @Input() public operator: "or" | "and" = "or";
  @Input() public limit: number | string = 10;
  @Input() public showMoreLimit: number | string;
  @Input() public sortBy: string[] | ((item: object) => number);

  public state = {
    canRefine: false,
    canToggleShowMore: false,
    createURL: noop,
    isShowingMore: false,
    items: [],
    refine: noop,
    toggleShowMore: noop
  };

  constructor(searchInstance: NgAisInstance) {
    super(searchInstance);
  }

  public ngOnInit() {
    // limit & showMoreLimit can be interferred as string from props
    // so we need to transform them back to number
    const limit =
      typeof this.limit === "string" ? parseInt(this.limit, 10) : this.limit;

    const showMoreLimit =
      typeof this.showMoreLimit === "string"
        ? parseInt(this.showMoreLimit, 10)
        : this.showMoreLimit;

    this.createWidget(connectRefinementList, {
      limit,
      showMoreLimit,
      attributeName: this.attributeName,
      sortBy: this.sortBy
    });

    super.ngOnInit();
  }

  public refine(event, item) {
    event.preventDefault();
    event.stopPropagation();

    if (this.state.canRefine) {
      // update UI directly, it will update the checkbox state
      item.isRefined = !item.isRefined;

      // refine through Algolia API
      this.state.refine(item.value);
    }
  }
}
