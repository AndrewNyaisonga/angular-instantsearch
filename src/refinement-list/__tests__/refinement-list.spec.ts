import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../../instantsearch/instantsearch.module";
import { NgAisRefinementListModule } from "../refinement-list.module";
import { NgAisRefinementList } from "../refinement-list";

jest.mock("../../base-widget");

const defaultState = {
  canRefine: true,
  canToggleShowMore: false,
  createURL: jest.fn(),
  isShowingMore: false,
  items: [
    { label: "foo", count: 100, value: "foo", isRefined: false },
    { label: "bar", count: 100, value: "bar", isRefined: false },
    { label: "foobar", count: 100, value: "foobar", isRefined: false },
    { label: "barfoo", count: 100, value: "barfoo", isRefined: false }
  ],
  refine: jest.fn(),
  toggleShowMore: jest.fn()
};

const render = (state?: {}) => {
  const fixture = TestBed.createComponent(NgAisRefinementList);

  if (state) {
    fixture.componentInstance.updateState({ ...defaultState, ...state }, false);
  }

  fixture.detectChanges();
  return fixture;
};

describe("RefinementList", () => {
  beforeEach(() =>
    TestBed.configureTestingModule({
      declarations: [],
      imports: [NgAisInstantSearchModule.forRoot(), NgAisRefinementListModule]
    })
  );

  it("renders markup without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("renders markup with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("should call `toggleShowMore()` on click", () => {
    const toggleShowMore = jest.fn();
    const fixture = render({ toggleShowMore, canToggleShowMore: true });

    const showMoreBtn = fixture.debugElement.nativeElement.querySelector(
      "button"
    );
    showMoreBtn.click();

    expect(toggleShowMore).toHaveBeenCalled();
  });

  it("should call `refine()` on item click", () => {
    const refine = jest.fn();
    const fixture = render({ refine });

    const [, secondItem] = fixture.debugElement.nativeElement.querySelectorAll(
      "li"
    );
    secondItem.click();

    expect(refine).toHaveBeenCalled();
    expect(refine).toHaveBeenCalledWith(defaultState.items[1].value);
  });
});
