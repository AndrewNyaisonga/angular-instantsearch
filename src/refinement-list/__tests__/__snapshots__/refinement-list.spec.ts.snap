// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RefinementList renders markup with state 1`] = `
<ng-ais-refinement-list
  limitMin={[Function Number]}
  operator={[Function String]}
  searchInstance={[Function NgAisInstance]}
  searchPlaceholder={[Function String]}
  searchQuery=""
  showLessLabel={[Function String]}
  showMoreLabel={[Function String]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-RefinementList"
  >
    
        
    <ng-ais-header
      classname="ais-RefinementList-header ais-header"
      ng-reflect-class-name="ais-RefinementList-header ais-"
    >
      
      
      
      
    
    </ng-ais-header>
    
  
        
    <div
      class="ais-RefinementList-body ais-body"
    >
      
          
      
      
  
          
      <ul
        class="ais-RefinementList-list"
      >
        
            
        
        <li
          class="ais-RefinementList-item"
          ng-reflect-ng-class="[object Object]"
        >
          
              
          <label
            class="ais-RefinementList-label"
          >
            
                
            <input
              class="ais-RefinementList-checkbox"
              type="checkbox"
            />
            
                foo
                
            <span
              class="ais-RefinementList-count"
            >
              
                  100
                
            </span>
            
              
          </label>
          
            
        </li>
        <li
          class="ais-RefinementList-item"
          ng-reflect-ng-class="[object Object]"
        >
          
              
          <label
            class="ais-RefinementList-label"
          >
            
                
            <input
              class="ais-RefinementList-checkbox"
              type="checkbox"
            />
            
                bar
                
            <span
              class="ais-RefinementList-count"
            >
              
                  100
                
            </span>
            
              
          </label>
          
            
        </li>
        <li
          class="ais-RefinementList-item"
          ng-reflect-ng-class="[object Object]"
        >
          
              
          <label
            class="ais-RefinementList-label"
          >
            
                
            <input
              class="ais-RefinementList-checkbox"
              type="checkbox"
            />
            
                foobar
                
            <span
              class="ais-RefinementList-count"
            >
              
                  100
                
            </span>
            
              
          </label>
          
            
        </li>
        <li
          class="ais-RefinementList-item"
          ng-reflect-ng-class="[object Object]"
        >
          
              
          <label
            class="ais-RefinementList-label"
          >
            
                
            <input
              class="ais-RefinementList-checkbox"
              type="checkbox"
            />
            
                barfoo
                
            <span
              class="ais-RefinementList-count"
            >
              
                  100
                
            </span>
            
              
          </label>
          
            
        </li>
        
          
      </ul>
      
  
          
      
      
        
    </div>
    
  
        
    <ng-ais-footer
      classname="ais-RefinementList-footer ais-footer"
      ng-reflect-class-name="ais-RefinementList-footer ais-"
    >
      
      
      
      
    
    </ng-ais-footer>
    
      
  </div>
    
</ng-ais-refinement-list>
`;

exports[`RefinementList renders markup without state 1`] = `
<ng-ais-refinement-list
  limitMin={[Function Number]}
  operator={[Function String]}
  searchInstance={[Function NgAisInstance]}
  searchPlaceholder={[Function String]}
  searchQuery=""
  showLessLabel={[Function String]}
  showMoreLabel={[Function String]}
  state={[Function Object]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-RefinementList"
  >
    
        
    <ng-ais-header
      classname="ais-RefinementList-header ais-header"
      ng-reflect-class-name="ais-RefinementList-header ais-"
    >
      
      
      
      
    
    </ng-ais-header>
    
  
        
    <div
      class="ais-RefinementList-body ais-body"
    >
      
          
      
      
  
          
      <ul
        class="ais-RefinementList-list"
      >
        
            
        
        
          
      </ul>
      
  
          
      
      
        
    </div>
    
  
        
    <ng-ais-footer
      classname="ais-RefinementList-footer ais-footer"
      ng-reflect-class-name="ais-RefinementList-footer ais-"
    >
      
      
      
      
    
    </ng-ais-footer>
    
      
  </div>
    
</ng-ais-refinement-list>
`;

exports[`RefinementList should apply \`transformItems\` if specified 1`] = `
<ng-ais-refinement-list
  limitMin={[Function Number]}
  operator={[Function String]}
  searchInstance={[Function NgAisInstance]}
  searchPlaceholder={[Function String]}
  searchQuery=""
  showLessLabel={[Function String]}
  showMoreLabel={[Function String]}
  state={[Function Object]}
  transformItems={[Function Function]}
  updateState={[Function Function]}
>
  
      <div
    class="ais-RefinementList"
  >
    
        
    <ng-ais-header
      classname="ais-RefinementList-header ais-header"
      ng-reflect-class-name="ais-RefinementList-header ais-"
    >
      
      
      
      
    
    </ng-ais-header>
    
  
        
    <div
      class="ais-RefinementList-body ais-body"
    >
      
          
      
      
  
          
      <ul
        class="ais-RefinementList-list"
      >
        
            
        
        <li
          class="ais-RefinementList-item"
          ng-reflect-ng-class="[object Object]"
        >
          
              
          <label
            class="ais-RefinementList-label"
          >
            
                
            <input
              class="ais-RefinementList-checkbox"
              type="checkbox"
            />
            
                transformed - foo
                
            <span
              class="ais-RefinementList-count"
            >
              
                  100
                
            </span>
            
              
          </label>
          
            
        </li>
        <li
          class="ais-RefinementList-item ais-RefinementList-item--selected"
          ng-reflect-ng-class="[object Object]"
        >
          
              
          <label
            class="ais-RefinementList-label"
          >
            
                
            <input
              class="ais-RefinementList-checkbox"
              type="checkbox"
            />
            
                transformed - bar
                
            <span
              class="ais-RefinementList-count"
            >
              
                  100
                
            </span>
            
              
          </label>
          
            
        </li>
        <li
          class="ais-RefinementList-item"
          ng-reflect-ng-class="[object Object]"
        >
          
              
          <label
            class="ais-RefinementList-label"
          >
            
                
            <input
              class="ais-RefinementList-checkbox"
              type="checkbox"
            />
            
                transformed - foobar
                
            <span
              class="ais-RefinementList-count"
            >
              
                  100
                
            </span>
            
              
          </label>
          
            
        </li>
        <li
          class="ais-RefinementList-item"
          ng-reflect-ng-class="[object Object]"
        >
          
              
          <label
            class="ais-RefinementList-label"
          >
            
                
            <input
              class="ais-RefinementList-checkbox"
              type="checkbox"
            />
            
                transformed - barfoo
                
            <span
              class="ais-RefinementList-count"
            >
              
                  100
                
            </span>
            
              
          </label>
          
            
        </li>
        
          
      </ul>
      
  
          
      
      
        
    </div>
    
  
        
    <ng-ais-footer
      classname="ais-RefinementList-footer ais-footer"
      ng-reflect-class-name="ais-RefinementList-footer ais-"
    >
      
      
      
      
    
    </ng-ais-footer>
    
      
  </div>
    
</ng-ais-refinement-list>
`;
